### PR TITLE
Add arrow left/right navigation to tree table

### LINKED ISSUE #1769: Support arrow key traversal in Tree Tables
This is implemented in the dart:html app. We likely need to add a "selected" state to the tree table items so that we have a modifiable starting point of traversal.

### PR DESCRIPTION
**Pull request by gmoothart**
Left key collapses the current node or selects the node's parent.
Right key expands the current node, or selects its first child.

Also fixed a math error in scrolling and added comments plus general
cleanup.

Fixes  #1769

### CODE REVIEW

File: packages/devtools_app/lib/src/flutter/table.dart [hunk 1]
```diff
@@ -1,4 +1,5 @@
 import 'dart:math';
+import 'dart:ui';
```
Comment (jacob314) [inline packages/devtools_app/lib/src/flutter/table.dart:2]: nit: by convention typically `import 'dart:ui' as ui` as it has some names that can overlap with other libraries we use.

File: packages/devtools_app/lib/src/flutter/table.dart [hunk 2]
```diff
@@ -446,36 +453,59 @@ class TreeTableState<T extends TreeNode<T>> extends State<TreeTable<T>>
...
+    else if (event.logicalKey == LogicalKeyboardKey.arrowLeft) {
+      if (selectedNode.isExpanded) {
+        _toggleNode(selectedNode);
+      } else {
+        _moveSelection(
+            ScrollKind.parent, scrollController, constraints.maxHeight);
+      }
+    }
+
+    // On right arrow expand the row if possible, otherwise move selection down.
```
Comment (jacob314) [inline packages/devtools_app/lib/src/flutter/table.dart:473]: good comments!

File: packages/devtools_app/lib/src/flutter/table.dart [hunk 3]
```diff
@@ -446,36 +453,59 @@ class TreeTableState<T extends TreeNode<T>> extends State<TreeTable<T>>
     }
     assert(selectedNode == items[selectedNodeIndex]);
 
-    if (event.logicalKey == LogicalKeyboardKey.arrowDown ||
-        event.logicalKey == LogicalKeyboardKey.arrowUp) {
-      _moveSelection(event.logicalKey, scrollController, constraints.maxHeight);
+    if (event.logicalKey == LogicalKeyboardKey.arrowDown) {
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/flutter/table.dart:456]: let's use a switch statement instead 
`switch (event.logicalKey) { ... }`

File: packages/devtools_app/lib/src/flutter/table.dart [hunk 4]
```diff
@@ -446,36 +453,59 @@ class TreeTableState<T extends TreeNode<T>> extends State<TreeTable<T>>
...
+    // '-1' in the following calculations is needed because the header row
+    // occupies space in the viewport so we must subtract it out.
     final minCompleteItemsInView =
         (viewportHeight / _Table.defaultRowHeight).floor() - 1;
     final lastItemIndex = firstItemIndex + minCompleteItemsInView - 1;
     int newSelectedNodeIndex;
 
-    if (key == LogicalKeyboardKey.arrowDown) {
-      // move selection down if there is a node lower down.
+    if (scrollKind == ScrollKind.down) {
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/flutter/table.dart:500]: could use a switch statement here as well to make it clear that there are three possible cases

File: packages/devtools_app/lib/src/flutter/table.dart [hunk 5]
```diff
@@ -446,36 +453,59 @@ class TreeTableState<T extends TreeNode<T>> extends State<TreeTable<T>>
...
+    }
+
+    // On left arrow collapse the row if it is expanded. If it is not, move
+    // selection to its parent.
+    else if (event.logicalKey == LogicalKeyboardKey.arrowLeft) {
+      if (selectedNode.isExpanded) {
+        _toggleNode(selectedNode);
+      } else {
+        _moveSelection(
+            ScrollKind.parent, scrollController, constraints.maxHeight);
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/flutter/table.dart:469]: trailing comma

File: packages/devtools_app/lib/src/flutter/table.dart [hunk 6]
```diff
@@ -446,36 +453,59 @@ class TreeTableState<T extends TreeNode<T>> extends State<TreeTable<T>>
...
+      }
+    }
+
+    // On right arrow expand the row if possible, otherwise move selection down.
+    else if (event.logicalKey == LogicalKeyboardKey.arrowRight) {
+      if (!selectedNode.isExpanded) {
+        _toggleNode(selectedNode);
+      } else {
+        _moveSelection(
+            ScrollKind.down, scrollController, constraints.maxHeight);
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/flutter/table.dart:479]: trailing comma

File: packages/devtools_app/lib/src/flutter/table.dart (hunk 3 above)
Comment (gmoothart) [reply packages/devtools_app/lib/src/flutter/table.dart:456]: I actually tried to do this, and got an error: `The switch case expression type 'LogicalKeyboardKey' can't override the == operator.`

LMK if there is a way to get around this.

Comment (kenzieschmoll) [reply packages/devtools_app/lib/src/flutter/table.dart:456]: strange. not a blocker feel free to land without this change

File: packages/devtools_app/lib/src/flutter/table.dart (hunk 1 above)
Comment (gmoothart) [reply packages/devtools_app/lib/src/flutter/table.dart]: It turns out this line wasn't necessary.

File: packages/devtools_app/lib/src/flutter/table.dart (hunk 3 above)
Comment (gmoothart) [reply packages/devtools_app/lib/src/flutter/table.dart:455]: I moved things around to chain the if/else blocks more obviously.

File: packages/devtools_app/lib/src/flutter/table.dart (hunk 4 above)
Comment (gmoothart) [reply packages/devtools_app/lib/src/flutter/table.dart]: Done.

File: packages/devtools_app/lib/src/flutter/table.dart (hunk 5 above)
Comment (gmoothart) [reply packages/devtools_app/lib/src/flutter/table.dart]: Done

File: packages/devtools_app/lib/src/flutter/table.dart (hunk 6 above)
Comment (gmoothart) [reply packages/devtools_app/lib/src/flutter/table.dart]: Done

**Pull request merged**